File: wavefront/client/src/api/app-user-service.ts
import { IApiResponse } from '@app/lib/axios';
import { IUser } from '@app/types/user';
import { AxiosInstance } from 'axios';

export class AppUserService {
  constructor(private http: AxiosInstance) {}

  /**
   * Grant user access to an app (owners only)
   */
  async grantAppAccess(appId: string, userId: string): Promise<IApiResponse<{ message: string }>> {
    return this.http.post(`/v1/apps/${appId}/users/${userId}`);
  }

  /**
   * Revoke user access from an app (owners only)
   */
  async revokeAppAccess(appId: string, userId: string): Promise<IApiResponse<{ message: string }>> {
    return this.http.delete(`/v1/apps/${appId}/users/${userId}`);
  }

  /**
   * List users with access to an app (owners only)
   */
  async listAppUsers(appId: string): Promise<IApiResponse<{ users: IUser[] }>> {
    return this.http.get(`/v1/apps/${appId}/users`);
  }

  /**
   * List apps accessible to a user (owners only)
   */
  async listUserApps(userId: string): Promise<IApiResponse<{ app_ids: string[] }>> {
    return this.http.get(`/v1/users/${userId}/apps`);
  }
}

File: wavefront/server/apps/floconsole/floconsole/services/app_user_service.py
from typing import List
from uuid import UUID

from floconsole.db.models.app_user import AppUser
from floconsole.db.repositories.sql_alchemy_repository import SQLAlchemyRepository


class AppUserService:
    """Service for managing app-user access relationships"""

    def __init__(self, app_user_repository: SQLAlchemyRepository[AppUser]):
        self.app_user_repository = app_user_repository

    async def grant_app_access(self, user_id: UUID, app_id: UUID) -> AppUser:
        """Grant user access to app"""
        # Use upsert to avoid duplicate key errors
        await self.app_user_repository.upsert(
            filters={'user_id': user_id, 'app_id': app_id}
        )

        # Return the created/existing record
        return await self.app_user_repository.find_one(user_id=user_id, app_id=app_id)

    async def revoke_app_access(self, user_id: UUID, app_id: UUID) -> bool:
        """Revoke user access to app"""
        await self.app_user_repository.delete_all(user_id=user_id, app_id=app_id)
        return True

    async def get_user_apps(self, user_id: UUID) -> List[AppUser]:
        """Get all apps a user has access to"""
        return await self.app_user_repository.find(user_id=user_id)

    async def get_app_users(self, app_id: UUID) -> List[AppUser]:
        """Get all users who have access to an app"""
        return await self.app_user_repository.find(app_id=app_id)
